### PR TITLE
KI-85 [tfjs] Conv operator

### PR DESCRIPTION
**Pull request by dmitriyb**

### CODE REVIEW

File: ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt [hunk 1]
```diff
@@ -514,3 +514,84 @@ suspend fun NumberNDArrayTFJS.batchNorm(
...
+        3 -> intArrayOf(2, 1, 0) to { i: ArrayTFJS, f: ArrayTFJS, s: Array<Number>, p: Any, d: Array<Number> ->
+            conv1d(i, f, s.first(), p, "NCW", d.first())
+        }
+        4 -> intArrayOf(2, 3, 1, 0) to { i: ArrayTFJS, f: ArrayTFJS, s: Array<Number>, p: Any, d: Array<Number> ->
+            conv2d(i, f, s, p, "NCHW", d)
+        }
+        5 -> intArrayOf(2, 3, 4, 1, 0) to { i: ArrayTFJS, f: ArrayTFJS, s: Array<Number>, p: Any, d: Array<Number> ->
+            conv3d(i, f, s, p, "NCDHW", d)
+        }
+        else -> error("Unsupported tensor rank for convolution: $rank")
```
Comment (cupertank) [inline ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt:544]: I think we need to specify which ranks are supported in this message

File: ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt [hunk 2]
```diff
@@ -514,3 +514,84 @@ suspend fun NumberNDArrayTFJS.batchNorm(
...
+        for (group in 0 until groups) {
+            // Calculate the slicing indices for input and filter
+            val inputStarts = IntArray(rank) { if (it == 1) group * channelsPerGroup else 0 }
+            val inputEnds = IntArray(rank) { if (it == 1) (group + 1) * channelsPerGroup else shape[it] }
+
+            val filterStarts = IntArray(rank) { if (it == 0) group * (w.shape[0] / groups) else 0 }
+            val filterEnds = IntArray(rank) { if (it == 0) (group + 1) * (w.shape[0] / groups) else w.shape[it] }
+
+            val inputSlice = this.slice(starts = inputStarts, ends = inputEnds).tfjsArray
+            val filterSlice = w.slice(starts = filterStarts, ends = filterEnds).transpose(transposeArr).tfjsArray
```
Comment (cupertank) [inline ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt:557]: Can we transpose weights out of cycle? It's pretty expensive operation

File: inference/inference-tfjs/src/jsMain/kotlin/io.kinference.tfjs/operators/conv/Conv.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    inputs: List<String>,
+    outputs: List<String>
+) : Operator<TFJSTensor, TFJSTensor>(name, info, attributes, inputs, outputs) {
+    companion object {
+        private val DEFAULT_VERSION = VersionInfo(sinceVersion = 11)
+
+        operator fun invoke(name: String, version: Int?, attributes: Map<String, Attribute<Any>>, inputs: List<String>, outputs: List<String>) =
+            when (version ?: DEFAULT_VERSION.sinceVersion) {
+                in ConvVer11.VERSION.asRange() -> ConvVer11(name, attributes, inputs, outputs)
+                else -> error("Unsupported version of Conv1d operator: $version")
```
[the remark this answers is not in the document]
Comment (dmitriyb) [reply inference/inference-tfjs/src/jsMain/kotlin/io.kinference.tfjs/operators/conv/Conv.kt]: Fixed in [2cd4c8e](https://github.com/JetBrains-Research/kinference/pull/160/commits/2cd4c8e2b833a65e64cece8aaf487e8636f0bdeb)

File: ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt [hunk 4]
```diff
@@ -514,3 +514,84 @@ suspend fun NumberNDArrayTFJS.batchNorm(
...
+    w: NumberNDArrayTFJS,
+    b: NumberNDArrayTFJS?,
+    groups: Int,
+    autoPad: String,
+    pads: IntArray?,
+    strides: IntArray?,
+    dilations: IntArray?
+) : NumberNDArrayTFJS {
+    val effectiveStrides: Array<Number> = (strides ?: IntArray(rank - 2) { 1 }).map { it }.toTypedArray()
+    val effectiveDilations: Array<Number> = (dilations ?: IntArray(rank - 2) { 1 }).map { it }.toTypedArray()
```
[the remark this answers is not in the document]
Comment (dmitriyb) [reply ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt]: Fixed in [2cd4c8e](https://github.com/JetBrains-Research/kinference/pull/160/commits/2cd4c8e2b833a65e64cece8aaf487e8636f0bdeb) with introducing ConvParameters class where all additional work with pads, strides, and dilations performs

File: ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt [hunk 5]
```diff
@@ -514,3 +514,84 @@ suspend fun NumberNDArrayTFJS.batchNorm(
...
+        } else
+            NumberNDArrayTFJS(y)
+    }
+}
+
+private fun convertOnnxPaddingToTfjs(autoPad: String, pads: IntArray?, numSpatialDims: Int): Any {
+    return when (autoPad) {
+        "SAME_UPPER", "SAME_LOWER" -> "same"
+        "VALID" -> "valid"
+        "NOTSET" -> {
```
[the remark this answers is not in the document]
Comment (dmitriyb) [reply ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt]: Fixed in [2cd4c8e](https://github.com/JetBrains-Research/kinference/pull/160/commits/2cd4c8e2b833a65e64cece8aaf487e8636f0bdeb) with introducing ConvParameters class where all additional work with pads, strides, and dilations performs

File: ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt (hunk 1 above)
Comment (dmitriyb) [reply ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt]: Fixed in [2cd4c8e](https://github.com/JetBrains-Research/kinference/pull/160/commits/2cd4c8e2b833a65e64cece8aaf487e8636f0bdeb)

File: ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt (hunk 2 above)
Comment (dmitriyb) [reply ndarray/ndarray-tfjs/src/jsMain/kotlin/io.kinference.ndarray/extensions/NDArrayExtension.kt]: Fixed in [2cd4c8e](https://github.com/JetBrains-Research/kinference/pull/160/commits/2cd4c8e2b833a65e64cece8aaf487e8636f0bdeb) with rewriting NumberNDArrayTFJS.conv: filter transposing is performed one time before group cycle

**Review by cupertank — APPROVED** [review]

**Pull request merged**